socks5.c

```diff
@@ -0,0 +1,337 @@
+    if(dill_slow(h < 0)) {err = errno; goto error;}
+    err = dill_socks5_clientauth(h, username, password, deadline);
+    if (err) {s = h; err = errno; goto error;}
+    return h;
+error:
+    if(s >= 0) dill_hclose(s);
+    errno = err;
+    return -1;
+}
+
+int dill_socks5_attach_client(int s, const char *username,
+      const char *password, int64_t deadline) {
+    int err;
+    struct dill_socks5_sock *obj = malloc(sizeof(struct dill_socks5_sock));
+    if(dill_slow(!obj)) {err = ENOMEM; goto error1;}
+    s = dill_socks5_attach_client_mem(s, username, password,
+        (struct dill_socks5_storage*)obj, deadline);
+    if(dill_slow(s < 0)) {err = errno; goto error2;}
+    obj->mem = 0;
+    return s;
+error2:
+    free(obj);
+error1:
+    if(s >= 0) dill_hclose(s);
+    errno = err;
+    return -1;
+}
+
+int dill_socks5_connect(int s, char *addr, int port, int64_t deadline) {
+    // validate socket type
+    struct dill_socks5_sock *obj = dill_hquery(s, dill_socks5_type);
+    if(dill_slow(!obj)) return -1;
+    if(dill_slow(obj->server)) {errno = EPROTO; return -1;}
+    if(dill_slow(obj->connect)) {errno = EISCONN; return -1;}
+    // validate input
+    if(dill_slow(!addr)) {errno = EINVAL; return -1;}
+    if(dill_slow(strlen(addr) > 255)) {errno = EINVAL; return -1;}
+    if((port <= 0) || (port>65535)) {errno = EINVAL; return -1;}
+    // largest possible connect request = 255 chars for name + 7 bytes for
+    // VER, CMD, RSV, ATYP, ALEN, PORT[2]
+    uint8_t conn[262];
+    size_t conn_len;
+    conn[0] = 0x05; // VER = 5
+    conn[1] = 0x01; // CMD = CONNECT
+    conn[2] = 0x00; // RSV
+    struct in_addr ina4;
+    struct in6_addr ina6;
+    if (inet_pton(AF_INET, addr, &ina4) == 1) {
+        // IPv4 Address, now stored in ina4, Network Byte Order
+        conn[3] = 0x01; // ATYP = IPV4
+        memcpy((void *)(conn + 4), (void *)&ina4, 4);
+        conn[8] = port >> 8;
+        conn[9] = port & 0xFF;
+        conn_len = 10;
+    } else if (inet_pton(AF_INET6, addr, &ina6) == 1) {
+        // IPv6 Address, now stored in ina6, Network Byte Order
+        conn[3] = 0x04; // ATYP = IPV6
+        memcpy((void *)(conn + 4), (void *)&ina6, 16);
+        conn[20] = port >> 8;
+        conn[21] = port & 0xFF;
+        conn_len = 22;
+   } else {
+        // assume a resolvable domain name
+        // previously validate to be 0 < x <= 255 bytes
+        uint8_t alen = (uint8_t)strlen(addr);
+        conn[3] = 0x03; // ATYP = DOMAINNAME
+        conn[4] = alen;
+        memcpy((void *)(conn+ 5), (void *)addr, alen);
+        conn[5+alen] = port >> 8;
+        conn[6+alen] = port & 0xFF;
+        conn_len = 7 + alen;
+    }
+    int err = dill_bsend(obj->u, conn, conn_len, deadline);
+    if (dill_slow(err)) {return -1;}
+    // reply has the same form/length limit, reuse request as reply buffer
+    // read first 5 octets (through ATYP) to determine how much more to read
+    err = dill_brecv(obj->u, conn, 5, deadline);
```

Comment:
You don't need to hyper-optimize here. brecv() does reading in batches, so each new brecv doesn't necessarily mean new user/kernel transition. Read 4 bytes, then 1 byte if you acually need it.

Reply:
In any case you have to read at least 5 bytes... the shortest option (IPv4 literal as 4 bytes) is 10 bytes in total.